pkg/internal/controller/metrics.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import "github.com/prometheus/client_golang/prometheus"
+
+// Metrics holds the prometheus metrics used internally by the controller
+type Metrics struct {
```

Comment:
I was considering the fact that each controller runs in a separate go-routine, perhaps it is better to have one instance of each metric per controller to reduce cross thread communication.

Happy to make the change as you've suggested if you're confident that writing to prometheus metrics is thread-safe? I wasn't sure whether they were or not so went for what I thought was the safer option

Reply:
yeah, Observe and friends are threadsafe:

> All exported functions and methods are safe to be used concurrently unless specified otherwise.

https://godoc.org/github.com/prometheus/client_golang/prometheus